File: backend/src/workers/tag-updater.ts
import { and, eq, gte, isNull, lt, or } from 'drizzle-orm';
import { db } from '../db';
import { tagHistory, tags } from '../db/schema';
import { fanslyClient } from '../fansly/client';
import { logger } from '../utils/logger';
import type { Worker } from './manager';

class TagUpdaterWorker implements Worker {
  name = 'tag-updater';
  interval: number;

  constructor() {
    this.interval = parseInt(process.env.WORKER_UPDATE_INTERVAL || '10000'); // Default: 10 seconds for continuous running
  }

  async run(): Promise<void> {
    logger.info('Starting tag update process');
    const rateLimitDelay = 60000 / parseInt(process.env.FANSLY_API_RATE_LIMIT || '60'); // Default: 60 requests per minute
    const twentyFourHoursAgo = new Date(Date.now() - 24 * 60 * 60 * 1000);

    try {
      // Fetch tags that haven't been updated in the last 24 hours
      const trackedTags = await db
        .select()
        .from(tags)
        .where(or(lt(tags.lastCheckedAt, twentyFourHoursAgo), isNull(tags.lastCheckedAt)))
        .limit(10); // Process in batches to avoid long-running operations

      if (trackedTags.length === 0) {
        logger.info('No tags need updating at this time');
        return;
      }

      logger.info(`Found ${trackedTags.length} tags that need updating`);

      let updated = 0;
      let errors = 0;

      for (const tag of trackedTags) {
        try {
          // Check if we already have data for today
          const todayStart = new Date();
          todayStart.setHours(0, 0, 0, 0);

          const existingTodayEntry = await db
            .select()
            .from(tagHistory)
            .where(and(eq(tagHistory.tagId, tag.id), gte(tagHistory.createdAt, todayStart)))
            .limit(1);

          if (existingTodayEntry.length > 0) {
            logger.info(`Tag "${tag.tag}" already has data for today, skipping`);

            // Update lastCheckedAt to prevent checking again today
            await db.update(tags).set({ lastCheckedAt: new Date() }).where(eq(tags.id, tag.id));

            continue;
          }

          // Fetch updated data from Fansly
          const tagData = await fanslyClient.getTag(tag.tag);

          if (tagData) {
            const currentViewCount = tagData.viewCount;
            const previousViewCount = tag.viewCount;

            // Update tag data
            await db
              .update(tags)
              .set({
                viewCount: currentViewCount,
                lastCheckedAt: new Date(),
              })
              .where(eq(tags.id, tag.id));

            // Create history record
            await db.insert(tagHistory).values({
              tagId: tag.id,
              viewCount: currentViewCount,
              change: currentViewCount - previousViewCount,
            });

            logger.info(
              `Updated tag "${tag.tag}": ${previousViewCount} -> ${currentViewCount} (${currentViewCount > previousViewCount ? '+' : ''}${currentViewCount - previousViewCount})`
            );

            updated++;
          } else {
            logger.warn(`Failed to fetch data for tag: ${tag.tag}`);
            errors++;
          }

          // Rate limiting delay
          await new Promise((resolve) => setTimeout(resolve, rateLimitDelay));
        } catch (error) {
          logger.error(`Error updating tag "${tag.tag}":`, error);
          errors++;
        }
      }

      logger.info(`Tag update completed. Updated: ${updated}, Errors: ${errors}`);
    } catch (error) {
      logger.error('Tag update process failed:', error);
      throw error;
    }
  }
}

export const tagUpdaterWorker = new TagUpdaterWorker();

File: backend-go/workers/tag_updater.go
package workers

import (
	"context"
	"errors"
	"fmt"
	"ftoolbox/config"
	"ftoolbox/fansly"
	"ftoolbox/models"
	"time"

	"go.uber.org/zap"
	"gorm.io/gorm"
)

type TagUpdaterWorker struct {
	BaseWorker
	db     *gorm.DB
	client *fansly.Client
}

func NewTagUpdaterWorker(db *gorm.DB, cfg *config.Config, client *fansly.Client) *TagUpdaterWorker {
	interval := time.Duration(cfg.WorkerUpdateInterval) * time.Millisecond

	return &TagUpdaterWorker{
		BaseWorker: NewBaseWorker("tag-updater", interval),
		db:         db,
		client:     client,
	}
}

func (w *TagUpdaterWorker) Run(ctx context.Context) error {
	zap.L().Info("Running tag updater")

	// Get tags that need updating (haven't been checked in 24 hours)
	twentyFourHoursAgo := time.Now().Add(-24 * time.Hour)

	var tags []models.Tag
	if err := w.db.Where("last_checked_at IS NULL OR last_checked_at < ?", twentyFourHoursAgo).
		Order("view_count DESC").
		Limit(20).
		Find(&tags).Error; err != nil {
		return fmt.Errorf("failed to fetch tags: %w", err)
	}

	if len(tags) == 0 {
		zap.L().Debug("No tags need updating")
		return nil
	}

	zap.L().Info("Updating tags", zap.Int("count", len(tags)))

	for _, tag := range tags {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		if err := w.updateTag(ctx, &tag); err != nil {
			zap.L().Error("Failed to update tag",
				zap.String("tag", tag.Tag),
				zap.Error(err))

			// Update last_checked_at even on error to avoid immediate retries
			now := time.Now()
			tag.LastCheckedAt = &now
			if updateErr := w.db.Save(&tag).Error; updateErr != nil {
				zap.L().Error("Failed to update last checked time after error",
					zap.String("tag", tag.Tag),
					zap.Error(updateErr))
			}
			continue
		}
	}

	return nil
}

func (w *TagUpdaterWorker) updateTag(ctx context.Context, tag *models.Tag) error {
	// Fetch current view count from Fansly
	viewCount, err := w.client.GetTagWithContext(ctx, tag.Tag)
	if err != nil {
		// Check if tag no longer exists on Fansly
		if errors.Is(err, fansly.ErrTagNotFound) {
			// Mark tag as deleted
			now := time.Now()
			tag.LastCheckedAt = &now
			tag.UpdatedAt = now

			// Only update deletion fields if not already marked as deleted
			if !tag.IsDeleted {
				tag.IsDeleted = true
				tag.DeletedDetectedAt = &now

				zap.L().Info("Tag no longer exists on Fansly, marking as deleted",
					zap.String("tag", tag.Tag))
			}

			// Save the updated tag (no history entry for deleted tags)
			if err := w.db.Save(tag).Error; err != nil {
				return fmt.Errorf("failed to update deleted tag: %w", err)
			}

			return nil
		}
		return fmt.Errorf("failed to fetch view count: %w", err)
	}

	// If tag was previously deleted but now exists again, clear the deletion flag
	if tag.IsDeleted {
		tag.IsDeleted = false
		tag.DeletedDetectedAt = nil
		zap.L().Info("Tag exists again on Fansly, clearing deleted status",
			zap.String("tag", tag.Tag))
	}

	// Calculate changes
	viewCountChange := viewCount.MediaOfferSuggestionTag.ViewCount - tag.ViewCount

	// Start transaction
	tx := w.db.Begin()

	// Update tag
	now := time.Now()
	tag.ViewCount = viewCount.MediaOfferSuggestionTag.ViewCount
	tag.LastCheckedAt = &now
	tag.UpdatedAt = now

	if err := tx.Save(tag).Error; err != nil {
		tx.Rollback()
		return fmt.Errorf("failed to update tag: %w", err)
	}

	history := models.TagHistory{
		TagID:     tag.ID,
		ViewCount: viewCount.MediaOfferSuggestionTag.ViewCount,
		Change:    viewCountChange,
	}

	if err := tx.Create(&history).Error; err != nil {
		tx.Rollback()
		return fmt.Errorf("failed to create history: %w", err)
	}

	// Commit transaction
	if err := tx.Commit().Error; err != nil {
		return fmt.Errorf("failed to commit transaction: %w", err)
	}

	zap.L().Debug("Updated tag",
		zap.String("tag", tag.Tag),
		zap.Int64("viewCount", viewCount.MediaOfferSuggestionTag.ViewCount),
		zap.Int64("viewCountChange", viewCountChange))

	return nil
}
